tools/metamorpheus/macros.xml

```diff
@@ -0,0 +1,712 @@
+    </xml>
+    <xml name="input_databases">
+        <param name="databases" type="data" format="uniprotxml,fasta" multiple="true" label="Reference Protein Databases"/>
+    </xml>
+    <xml name="input_toml">
+        <param name="input_toml" type="data" format="toml" optional="true" label="Optional Config toml from history"/>
+    </xml>
+    <token name="@CMD_IMPORTS@">
+#import re
+#def identifier_or_name($input1)
+    #if hasattr($input1, 'element_identifier')
+        #return $input1.element_identifier
+    #else
+        #return $input1.name
+    #end if
+#end def
+#def clean($name1)
+    #set $name_clean = $re.sub('[^\w\-_]', '_', $re.sub('(?i)[.](fa|fasta|xml|imzml|mzml|raw|mgf)$','', $re.sub('.*/','', $name1.rstrip('.gz'))))
+    #return $name_clean
+#end def
+#def ln_name($ds)
+    #set $ext = ''
+    #if $ds.is_of_type('mzml') or $ds.is_of_type('imzml')
+        #set $ext = ".mzML"
+    #else if $ds.is_of_type('thermo.raw')
+        #set $ext = ".raw"
+    #else if $ds.is_of_type('toml')
+        #set $ext = ".toml"
+    #else if $ds.is_of_type('uniprotxml')
+        #set $ext = ".xml"
+    #else if $ds.is_of_type('fasta')
+        #set $ext = ".fasta"
+    #end if
+    #set $name = "%s%s" % ($clean($identifier_or_name($ds)),$ext)
+    #return $name
+#end def
+    </token>
+
+    <token name="@PREPARE_INPUTS@"><![CDATA[
+    @CMD_IMPORTS@
+    #set $spectrum_names = []
+    #set $database_names = []
+    #set $inputs_dir = 'inputs'
+    mkdir -p $inputs_dir &&
+    #for $sf in $spectrums
+      #set $i_name = $ln_name($sf)
+      $spectrum_names.append($i_name)
+      ln -s '${sf}' '${inputs_dir}/${i_name}' &&
+    #end for
+    #for $db in $databases
+      #set $d_name = $ln_name($db)
+      $database_names.append($d_name)
+      ln -s '${db}' '${d_name}' &&
+    #end for
+]]></token>
+    <token name="@LN_CALIB_TOML@"><![CDATA[
+    #for $tf in $calib_tomls
+      #set $t_name = $ln_name($tf)
+      ln -s '${tf}' '${inputs_dir}/${t_name}' &&
+    #end for
+]]></token>
+    
+    <token name="@MERGE_TOMLS@"><![CDATA[
+#def merge_tomls($tomls, $merged):
+    #import re
+    #set $config = dict()
+    #set $config['categories'] = []
+    #set $category = ' ' 
+    #for $toml in $tomls
+        #set $fh = open(toml,'r')
+        #for $line in $fh:
+            #set $entry = $line.rstrip()
+            #if $entry.strip() == '':
+                #continue
+            #end if
+            #set $m = $re.match('^((TaskType)\s*=\s*(.*))$', $entry)
+            #if $m:
+                #if $m.groups()[0] not in $config:
+                    #set $config['TaskType'] = $entry
+                    #continue
+                #end if 
+            #end if
+            #set $m = $re.match('^(\[.+\])$', $entry)
+            #if $m:
+                #set $category = $m.groups()[0]
+                #if $category not in $config['categories']:
```

Comment:
I think `#if $category not in $config` should work, i.e. it should not be necessary to store the categories in  `$config['categories']`.